src/Google.Api.Gax/Api/Gax/PathTemplate.cs

```diff
@@ -0,0 +1,439 @@
+        /// but unnamed parameters have a name of <c>null</c>.
+        /// </summary>
+        public ReadOnlyCollection<string> ParameterNames { get; }
+
+        /// <summary>
+        /// Constructs a template from its textual representation, such as <c>shelves/*/books/**</c>.
+        /// </summary>
+        /// <param name="template">The textual representation of the template.</param>
+        public PathTemplate(string template)
+        {
+            _segments = template.Split(SlashSplit).Select(Segment.Parse).ToList();
+            _parameterSegments = _segments.Where(s => s.Kind != SegmentKind.Literal).ToList();
+            int pathWildcardCount = _segments.Count(s => s.Kind == SegmentKind.PathWildcard);
+            if (pathWildcardCount > 1)
+            {
+                throw new ArgumentException("Template contains multiple path wildcards", nameof(template));
+            }
+            _hasPathWildcard = pathWildcardCount != 0;
+            _originalTemplate = template;
+            ParameterNames = _parameterSegments.Select(x => x.Value).ToList().AsReadOnly();
+        }
+
+        /// <summary>
+        /// The number of parameter segments (regular wildcards or path wildcards, named or unnamed) in the template.
+        /// </summary>
+        public int ParameterCount => _parameterSegments.Count;
+
+        /// <summary>
+        /// Validates a service name, ensuring it is not empty and doesn't contain any slashes.
+        /// (In the future, we may want to make this stricter, e.g. that it's a valid DNS-like name.)
+        /// </summary>
+        /// <param name="serviceName">The name to validate</param>
+        /// <param name="parameterName">The name of the parameter</param>
+        internal static void ValidateServiceName(string serviceName, string parameterName)
+        {
+            if (serviceName == null)
+            {
+                return;
+            }
+            if (serviceName == "")
+            {
+                throw new ArgumentException("Service name cannot be empty", parameterName);
+            }
+            if (serviceName.Contains("/"))
+            {
+                throw new ArgumentException("Service name cannot be contain /", parameterName);
+            }
+        }
+
+        /// <summary>
+        /// Validate a single value from a sequence. This is used in both parsing and instantiating.
+        /// </summary>
+        internal void ValidateResourceId(int index, string resourceId)
+        {
+            _parameterSegments[index].ValidateWildcardValue(resourceId);
+        }
+
+        /// <summary>
+        /// Validates a whole array of resource IDs, including that the count matches.
+        /// </summary>
+        internal void ValidateResourceIds(string[] resourceIds)
+        {
+            if (resourceIds.Length != ParameterCount)
+            {
+                throw new ArgumentException($"Expected {ParameterCount} ids, got {resourceIds.Length}");
```

Comment:
Done. Fortunately the calling public method always has a parameter called resourceIds :)